OpenRA.Mods.Common/Traits/DockOnDeploy.cs

```diff
@@ -0,0 +1,125 @@
+#region Copyright & License Information
+/*
+ * Copyright (c) The OpenRA Developers and Contributors
+ * This file is part of OpenRA, which is free software. It is made
+ * available to you under the terms of the GNU General Public License
+ * as published by the Free Software Foundation, either version 3 of
+ * the License, or (at your option) any later version. For more
+ * information, see COPYING.
+ */
+#endregion
+
+using System.Collections.Generic;
+using OpenRA.Activities;
+using OpenRA.Mods.Common.Activities;
+using OpenRA.Mods.Common.Orders;
+using OpenRA.Primitives;
+using OpenRA.Traits;
+
+namespace OpenRA.Mods.Common.Traits
+{
+	[Desc("Finds the closest dock on deploy.")]
+	public class DockOnDeployInfo : ConditionalTraitInfo, Requires<DockClientManagerInfo>
+	{
+		[Desc("Docking type.")]
+		public readonly BitSet<DockType> DockType;
+
+		[CursorReference]
+		[Desc("Cursor to display when able to (un)deploy the actor.")]
+		public readonly string DeployCursor = "deploy";
+
+		[CursorReference]
+		[Desc("Cursor to display when unable to (un)deploy the actor.")]
+		public readonly string DeployBlockedCursor = "deploy-blocked";
+
+		public override object Create(ActorInitializer init) { return new DockOnDeploy(init.Self, this); }
+	}
+
+	public class DockOnDeploy : ConditionalTrait<DockOnDeployInfo>, IIssueDeployOrder, IResolveOrder, IOrderVoice, IIssueOrder
+	{
+		readonly DockClientManager manager;
+		public DockOnDeploy(Actor self, DockOnDeployInfo info)
+			: base(info)
+		{
+			manager = self.Trait<DockClientManager>();
+		}
+
+		public IEnumerable<IOrderTargeter> Orders
+		{
+			get
+			{
+				yield return new DeployOrderTargeter("FindDock", 10,
+					() => !IsTraitDisabled && manager.DockingPossible(Info.DockType) ? Info.DeployCursor : Info.DeployBlockedCursor);
+			}
+		}
+
+		public Order IssueOrder(Actor self, IOrderTargeter order, in Target target, bool queued)
+		{
+			if (order.OrderID == "FindDock")
+				return new Order(order.OrderID, self, queued);
+
+			return null;
+		}
+
+		Order IIssueDeployOrder.IssueDeployOrder(Actor self, bool queued)
+		{
+			return new Order("FindDock", self, queued);
+		}
+
+		bool IIssueDeployOrder.CanIssueDeployOrder(Actor self, bool queued) { return !IsTraitDisabled && manager.DockingPossible(Info.DockType, queued); }
+
+		void IResolveOrder.ResolveOrder(Actor self, Order order)
+		{
+			if (order.OrderString == "FindDock" && !IsTraitDisabled && manager.DockingPossible(Info.DockType, order.Queued))
```

Comment:
This check cannot run at order resolving time! (Except the order string check.)

Reply:
fixed